helix-lsp/src/client.rs

```diff
@@ -36,11 +52,81 @@ pub struct Client {
     config: Option<Value>,
     root_path: std::path::PathBuf,
     root_uri: Option<lsp::Url>,
-    workspace_folders: Vec<lsp::WorkspaceFolder>,
+    workspace_folders: RwLock<Vec<lsp::WorkspaceFolder>>,
     req_timeout: u64,
 }
 
 impl Client {
+    pub fn try_add_doc(
+        &self,
+        root_markers: &[String],
+        manual_roots: &[PathBuf],
+        doc_path: Option<&std::path::PathBuf>,
+    ) -> bool {
+        // TODO how to handle this case
+        if !self.is_initialized() {
+            return true;
+        }
+        let (workspace, workspace_is_cwd) = find_workspace();
+        let root = find_lsp_workspace(
+            doc_path
+                .and_then(|x| x.parent().and_then(|x| x.to_str()))
+                .unwrap_or("."),
+            root_markers,
+            manual_roots,
+            &workspace,
+            workspace_is_cwd,
+        );
+        let root_uri = root.and_then(|root| lsp::Url::from_file_path(root).ok());
+        let mut workspace_folders = self.workspace_folders.write();
+        if self.root_uri == root_uri
+            || root_uri.as_ref().map_or(false, |root_uri| {
+                workspace_folders
+                    .iter()
+                    .any(|workspace| &workspace.uri == root_uri)
+            })
+        {
+            // workspace URI is already register we can use this client
+            return true;
+        }
+        if let Some(workspace_folders_caps) = self
+            .capabilities
+            .get()
+            .expect("Capabilities are known after initalization")
```

Comment:
Should we just stop and print an error if we don't have the capabilities, rather than crashing?

Reply:
I simply inlined the `capabilities` function here which does the same thing. The capabilities are always filled on initialization, currently I don't handle uninitialized LSPs (see the todo a bit earlier). I habe some ideas how to handle that more gracefully (related to the changes I am talking about here: https://github.com/helix-editor/helix/pull/5748#issuecomment-1421637983, I don't habe time to work in that right now but once I do, I will hopefully also treat preinitialization servers more gracefully.

Reply:
unitalzied LSPs are now handeled properly